Add cluster node sets

jhernand opened the pull request:
This patch adds to the `Cluster` type a new `node_sets` field to describe the sets of nodes that are part of the cluster, and to allow the user to change their size.

For example, a cluster with two node sets, one of them containing hosts with GPUs and another without could be represented like this:

```json
{
  "id": "123",
  "spec": {
    "node_sets": {
      "compute": {
        "class": "acme_1tb",
        "size": 3
      },
      "gpu": {
        "class": "acme_1tb_h100",
        "size": 1
      }
    }
  },
  "status": {
    "state": "CLUSTER_STATE_READY",
    "node_sets": {
      "compute": {
        "class": "acme_1tb",
        "size": 3
      },
      "gpu": {
        "class": "acme_1tb_h100",
        "size": 1
      }
    }
  }
}
```

The user will be able to scale each node set updating the `size` field of the `spec`.

The user will not have permission to edit any other field, or to add or remove node sets.

If the system is not able to resize the node set as requested it will set the `DEGRADED` condition to `true`, and will indicate the reason in the `message` field.

In order to provide information about the values of the `class` field of the node set the API will also have a new `HostClass` type and a `HostClases` service.

Related: https://github.com/innabox/issues/issues/99

tzumainn reviewed (commented):
It feels like ClusterNodeSet is a useful abstraction for (future) bare metal fulfillment as well; would it make sense to simply call it "NodeSet" to allow for more generic usage?

jhernand commented:
While working on a possible implementation for this I realized that representing the node sets as a list ins' t convenient, because there is no easy way in gRPC the update a field of a specific element of the list. The [field mask](https://protobuf.dev/reference/protobuf/google.protobuf/#field-mask) mechanism that is recommended for updates doesn't easily support array indexes.

To simplify that, both for the user and for the server, I have replaced the list with a map, where the key of the map is the identifier of the node set. For example:

```json
"node_sets": {
  "compute": {
    "class": "acme_1tb",
    "size": 3
  },
  "gpu": {
    "class": "acme_1tb_h100",
    "size": 1
  }
}
```

That simplifies things because it is easy to reference the `size` field of any of the node sets with a path like `spec.node_sets.compute.size` or `spec.node_sets.gpu.size`.

I have updated the code accordingly.

mhrivnak commented on proto/fulfillment/v1/cluster_type.proto:
```diff
@@ -33,9 +33,57 @@ message Cluster {
+  //   "id": "123",
+  //   "spec": {
+  //     "node_sets": {
+  //       "compute": {
```
In this case, is "compute" just an arbitrary name assigned to this particular node set?

If so, it would be much more k8s-idiomatic to make this an array of node sets, and make each one have an explicit "name" field. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

jhernand replied:
Yes, "compute" is a name assigned by the system to this node set.

What you suggest is what I did in the first iteration of this patch, but as I explained [here](https://github.com/innabox/fulfillment-api/pull/25#issuecomment-2820867281) protobuf an gRPC it complicates things because the recommended mechanism for that is the [field mask](https://protobuf.dev/reference/protobuf/google.protobuf/#field-mask) and that doesn't support updating individual elements of arrays.

We can anyhow use the array if we want, it will just be a bit more complicated to implement, as the server will need to learn that this specific fields needs special handling.

larsks replied:
I've opened https://github.com/innabox/fulfillment-api/issues/26 related to this line of questions.

mhrivnak commented on proto/fulfillment/v1/cluster_type.proto:
````diff
@@ -33,9 +33,57 @@ message Cluster {
+  // }
+  // ```
+  //
+  // The user will be allowed to update the `size` parameter, and the system will decide to allow or reject the request
````
By saying that the request will be allowed or rejected, that sounds like the API request itself may be denied. So if I try to go from size 2 to size 10, that could be rejected and the size would remain at 2. But if that was the behavior, it wouldn't really make sense to mark the cluster as degraded.

So what I think you might be describing is that the request will be accepted, and the spec will say 10. But then maybe the system can only get me up to 7 for some reason, and in that case the cluster will be degraded. If this is indeed the behavior you expect, I'd change the description to say something like "if at any time the system cannot allocate the number of nodes requested, the cluster will be marked as degraded with a reason recorded in the condition."

jhernand replied:
I meant the latter, will update the documentation.

mhrivnak commented on proto/fulfillment/v1/host_class_type.proto:
```diff
@@ -0,0 +1,41 @@
+// This is similar to the _instance type_ concept used by many cloud providers.
+//
+// The detailed chracteristics of the host (CPU, memory, GPU, etc) will be in the `description` field.
+message HostClass {
```
Does this have any direct relation to the ClusterNodeSet? How would this API get used? Is this a way to list the classes I can choose from, and if so, what value from here do I put into the ClusterNodeSet `class` field?

jhernand replied:
The relation is that the key of the map of node sets (or the `name` of the node set, if we use an array instead of a map) will be the identifier of one of these objects. So if the user sees a node set with key (or name) "xyz23" she can fetch the details of host class "xyz23" and read the description field to understand what that means.

The user can't set the  `class` field, it is read only.

In the future we may want to use this host class concept for bare metal provisioning, but for now it is just informative.

jhernand commented:
The new version of the patch renames `node_set.class` to `node_set.host_class` and improves the documentation.
